### PR TITLE
Create LnkExecutionTestSuite table

### PR DESCRIPTION
**Pull request by Cheshiriks**
- created table, entity, DTO

### CODE REVIEW

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt [hunk 1]
```diff
@@ -400,10 +402,8 @@ class TestSuitesSourceController(
     ): Mono<ByteBufferFluxResponse> = blockingToMono {
         val execution = executionService.findExecution(executionId)
             .orNotFound { "Execution (id=$executionId) not found" }
-        val testSuiteId = execution.parseAndGetTestSuiteIds()?.firstOrNull()
-            .orConflict { "Execution (id=$executionId) doesn't contain testSuiteIds" }
-        testSuitesService.findTestSuiteById(testSuiteId)
-            .orNotFound { "TestSuite (id=$testSuiteId) not found" }
+        val testSuite = lnkExecutionTestSuiteService.getAllTestSuitesByExecution(execution).first()
```
Comment (nulls) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt:405]:
```suggestion
        val testSuite = lnkExecutionTestSuiteService.getAllTestSuitesByExecution(execution).firstOrNull()
            .orConflict { "Execution (id=$executionId) doesn't have mapped test suites" }
```

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Execution.kt [hunk 2]
```diff
@@ -60,8 +59,6 @@ class Execution(
     @Enumerated(EnumType.STRING)
     var status: ExecutionStatus,
 
-    var testSuiteIds: String?,
```
Comment (nulls) [inline save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Execution.kt:63]: what's about `@ManyToMany` and add List of TestSuite here?

Comment (Cheshiriks) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Execution.kt:63]: I thought about it, but decided that it is not necessary. It is rarely used. And the LnkExecutionTestSuite does not store any information that could be passed to the executionDTO for transmission to the front.

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt [hunk 3]
```diff
@@ -52,9 +51,17 @@ class ExecutionService(
...
     /**
      * @param execution
-     * @return created/updated [Execution]
+     * @param testSuites
      */
-    fun saveExecution(execution: Execution): Execution = executionRepository.save(execution)
+    @Transactional
+    fun save(execution: Execution, testSuites: List<TestSuite>): Execution {
+        val newExecution = executionRepository.save(execution)
+        testSuites.map {
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt (hunk 1 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/TestSuitesSourceController.kt]: Fixed

File: buildSrc/src/main/kotlin/com/saveourtool/save/buildutils/DockerStackConfiguration.kt [hunk 4]
```diff
@@ -53,6 +53,7 @@ fun Project.createStackDeployTask(profile: String) {
                            |      - "3306:3306"
                            |    environment:
                            |      - "MYSQL_ROOT_PASSWORD=123"
+                           |    command: ["--log_bin_trust_function_creators=1"]
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply buildSrc/src/main/kotlin/com/saveourtool/save/buildutils/DockerStackConfiguration.kt:56]: MySQL is very limited in its capabilities. This option will allow you to create custom functions and use them. I originally wanted to use it for data migration. It is not used now, but I decided to keep it if it is useful in the future.

[the remark this answers is not in the document]
Comment (0x6675636b796f75676974687562) [reply buildSrc/src/main/kotlin/com/saveourtool/save/buildutils/DockerStackConfiguration.kt:56]: See [`log_bin_trust_function_creators`](https://dev.mysql.com/doc/refman/5.7/en/replication-options-binary-log.html#sysvar_log_bin_trust_function_creators)

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuite.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package com.saveourtool.save.entities
+
+import com.saveourtool.save.spring.entity.BaseEntityWithDto
+import javax.persistence.Entity
+import javax.persistence.JoinColumn
+import javax.persistence.ManyToOne
+
+/**
+ * @property execution
```
Comment (0x6675636b796f75676974687562) [inline save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuite.kt:9]: What's the point of this dummy KDoc comment?

To pacify `diktat` and make it happy?
But the corresponding `diktat` rule is enabled on purpose, I believe.

You should either disable the rule globally, or suppress it locally, or go ahead and add the proper KDoc comment.

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuiteDto.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package com.saveourtool.save.entities
+
+import com.saveourtool.save.execution.ExecutionDto
+import com.saveourtool.save.testsuite.TestSuiteDto
+import kotlinx.serialization.Serializable
+
+/**
+ * @property execution
```
Comment (0x6675636b796f75676974687562) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuiteDto.kt:8]:
```kotlin
@Suppress("KDOC_NO_CONSTRUCTOR_PROPERTY")
```

will express your intention more concisely.

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuite.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuite.kt]: Ok. Fixed KDoc comment

File: save-backend/src/test/kotlin/com/saveourtool/save/backend/controllers/RunExecutionControllerTest.kt [hunk 7]
```diff
@@ -164,12 +163,11 @@ class RunExecutionControllerTest(
 
         assertions.forEach { Assertions.assertNotNull(it) }
         val testsCount = testRepository.findAll()
-            .count { it.testSuite.requiredId() in originalExecution.parseAndGetTestSuiteIds().orEmpty() }
+            .count { it.testSuite.requiredId() in listOf(11L) }
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-backend/src/test/kotlin/com/saveourtool/save/backend/controllers/RunExecutionControllerTest.kt]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/LnkExecutionTestSuiteService.kt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Service of [LnkExecutionTestSuite]
+ */
+@Service
+class LnkExecutionTestSuiteService(
+    private val lnkExecutionTestSuiteRepository: LnkExecutionTestSuiteRepository,
+) {
+    /**
+     * @param execution
```
Comment (0x6675636b796f75676974687562) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/LnkExecutionTestSuiteService.kt:17]: Here and below: I suggest to either document the parameter, or add `@Suppress("KDOC_WITHOUT_PARAM_TAG")`.

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt [hunk 9]
```diff
@@ -52,9 +51,14 @@ class ExecutionService(
 
     /**
      * @param execution
-     * @return created/updated [Execution]
+     * @param testSuites
```
Comment (0x6675636b796f75676974687562) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt:54]: `@Suppress("KDOC_WITHOUT_PARAM_TAG")`

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/LnkExecutionTestSuiteRepository.kt [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import com.saveourtool.save.spring.repository.BaseEntityRepository
+import org.springframework.stereotype.Repository
+
+/**
+ * Repository of [LnkExecutionTestSuite]
+ */
+@Repository
+interface LnkExecutionTestSuiteRepository : BaseEntityRepository<LnkExecutionTestSuite> {
+    /**
+     * @param execution
```
Comment (0x6675636b796f75676974687562) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/LnkExecutionTestSuiteRepository.kt:14]: `@Suppress("KDOC_WITHOUT_PARAM_TAG")`

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/LnkExecutionTestSuiteService.kt (hunk 8 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/LnkExecutionTestSuiteService.kt]: Fixed KDoc comment

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt (hunk 9 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt]: Fixed KDoc comment

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/LnkExecutionTestSuiteRepository.kt (hunk 10 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/repository/LnkExecutionTestSuiteRepository.kt]: Fixed KDoc comment

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt [hunk 11]
```diff
@@ -214,25 +219,28 @@ class ExecutionService(
...
+            additionalFiles = execution.additionalFiles,
+            username = username,
+            sdk = execution.sdk,
+            execCmd = execution.execCmd,
+            batchSizeForAnalyzer = execution.batchSizeForAnalyzer,
+            testingType = execution.type,
+            contestName = lnkContestExecutionService.takeIf { execution.type == TestingType.CONTEST_MODE }
+                ?.findContestByExecution(execution)?.name,
+        )
+    }
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ExecutionService.kt:238]: I am not sure it is a good idea to implement reactive methods is such manner.

**Review by sanyavertolet — APPROVED** [review]
LGTM, but I would really love if someone else checked it. @petertrr

**Review by petertrr — APPROVED** [review]

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuiteDto.kt (hunk 6 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/entities/LnkExecutionTestSuiteDto.kt:8]: Fixed KDoc comment

**Pull request merged**